TraderJoe-Bots/Large-Banker-Joe-Transfers/src/agent.ts

```diff
@@ -0,0 +1,80 @@
+import { BigNumber } from "ethers";
+import { getEthersProvider } from "forta-agent";
+import { HandleTransaction, TransactionEvent, Finding } from "forta-agent";
+import { createFinding } from "./finding";
+import MarketsFetcher from "./markets.fetcher";
+import NetworkData from "./network";
+import SupplyFetcher from "./supply.fetcher";
+import { EVENTS_ABIS, MARKET_UPDATE_ABIS, PERCENTAGE } from "./utils";
+
+let networkManager = new NetworkData();
+let marketsFetcher = new MarketsFetcher(getEthersProvider());
+const supplyFetcher = new SupplyFetcher(getEthersProvider());
+
+const provideInitialize = (marketsFetcher: MarketsFetcher) => async () => {
+  const { chainId } = await marketsFetcher.provider.getNetwork();
+  networkManager.setNetwork(chainId);
+
+  // set joeTroller contract address
+  marketsFetcher.setJoeTrollerContract(networkManager.joeTroller);
+
+  // fetch Markets from joeTroller
+  await marketsFetcher.getMarkets("latest");
+};
+
+export const provideHandleTransaction =
+  (
+    networkManager: NetworkData,
+    marketsFetcher: MarketsFetcher,
+    supplyFetcher: SupplyFetcher
+  ): HandleTransaction =>
+  async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    // Update the markets list.
+    txEvent
+      .filterLog(MARKET_UPDATE_ABIS, networkManager.joeTroller)
+      .forEach((log) => {
+        marketsFetcher.updateMarkets(log.name, log.args.jToken.toString());
+      });
+
+    // Listen to `Mint`, `Redeem` and `Borrow` events on jToken contracts.
+    const logs = txEvent.filterLog(
+      EVENTS_ABIS,
+      Array.from(marketsFetcher.markets)
+    );
+    // fetch totalSupply for included markets.
+    const supplies = await Promise.all(
+      logs.map((log) =>
+        supplyFetcher.getTotalSupply(log.address, txEvent.blockNumber - 1)
+      )
+    );
+
+    for (let i = 0; i < logs.length; i++) {
```

Comment:
`supplies` is generated using `map` on logs so I don't see any risk of having a difference in length. 

The only scenario that might happen is a failed network call when fetching supplies, and that will cause the supply value to be undefined but the length will remain unchanged. We can catch this error inside the fetcher and set the supply to 0 without writing this value on cache, wdyt?

Reply:
You're right about the length being the same. So I guess that shouldn't be an issue.

In terms of catching the error and setting it to 0, I think that could work. However, I think we could set that to cache too, no? That way, if we're trying to fetch the same `totalSupply` at the same block for the same token, we could prevent a network call. In essence, treating it like a successful call to `totalSupply`. Thoughts?

Reply:
If we set the cache with the 0 value. Our bot won't try to fetch the correct `totalSupply` again and will always use a wrong value (`0`) for many transactions . 

I suggested catching the error but not setting cache just to make sure the bot won't crash on a failed network call. It will continue operating and will try making the call again in another transaction.

Reply:
I see. Makes sense to me to do it this way.